File: src/memory/paths.ts
/**
 * Auto-memory directory layout. Faithful port of Claude Code's `memdir/paths.ts`
 * (individual-only mode — no team memory, no Cowork/KAIROS overrides).
 *
 * Memory lives under the SAME per-project key the session store uses:
 *   ~/.deepdive/projects/<sanitized-cwd>/memory/
 *     MEMORY.md            — the index (loaded into context every session)
 *     <topic>.md           — one fact per file (frontmatter + body)
 *
 * The directory is pre-created each session (`ensureMemoryDirExists`) so the
 * model can write topic files without first running an `ls`/`mkdir`.
 */

import { mkdirSync } from "node:fs";
import { isAbsolute, join, normalize, sep } from "node:path";
import { getOriginalCwd } from "../workspace.js";
import { getProjectDir } from "../session.js";

export const MEMORY_DIRNAME = "memory";
export const ENTRYPOINT_NAME = "MEMORY.md";

/**
 * Whether auto-memory features (system-prompt section, index injection, recall,
 * background extraction) are active this session. Enabled by default; set
 * `DEEPDIVE_DISABLE_AUTO_MEMORY=1` to turn the whole subsystem off.
 */
export function isAutoMemoryEnabled(): boolean {
  const v = process.env.DEEPDIVE_DISABLE_AUTO_MEMORY;
  return !(v === "1" || v === "true");
}

/** The auto-memory directory for the current project (no trailing separator). */
export function getMemoryDir(): string {
  return join(getProjectDir(getOriginalCwd()), MEMORY_DIRNAME);
}

/** `<memoryDir>/MEMORY.md` — the index entrypoint loaded into context. */
export function getMemoryEntrypoint(): string {
  return join(getMemoryDir(), ENTRYPOINT_NAME);
}

/**
 * Ensure the memory directory exists. Idempotent, best-effort — called once per
 * session at prompt-build time so the model can Write to it directly. A real
 * failure (EACCES/EROFS) surfaces later on the model's own Write.
 */
export function ensureMemoryDirExists(): void {
  try {
    mkdirSync(getMemoryDir(), { recursive: true });
  } catch {
    // best-effort; the model's write_file does its own parent mkdir
  }
}

/**
 * Is `absolutePath` inside the auto-memory directory? Used by the permission
 * carve-out (memory writes/reads never prompt) and by the extraction guard
 * (which restricts the forked agent's writes to this directory).
 *
 * Normalizes first so `..` traversal can't slip a non-memory path past the
 * prefix check.
 */
export function isAutoMemPath(absolutePath: string): boolean {
  if (!absolutePath || !isAbsolute(absolutePath)) return false;
  const norm = normalize(absolutePath);
  const dir = normalize(getMemoryDir());
  return norm === dir || norm.startsWith(dir + sep);
}

File: deepdive-rs/crates/deepdive-core/src/memory/paths.rs
//! Auto-memory directory layout. Port of the TS `src/memory/paths.ts`
//! (individual-only mode). Memory lives under the SAME per-project key the
//! session store uses:
//!
//! ```text
//! ~/.deepdive/projects/<sanitized-cwd>/memory/
//!     MEMORY.md      — the index (loaded into context every session)
//!     <topic>.md     — one fact per file (frontmatter + body)
//! ```

use crate::session::get_project_dir;
use std::path::{Path, PathBuf};

pub const MEMORY_DIRNAME: &str = "memory";
pub const ENTRYPOINT_NAME: &str = "MEMORY.md";

/// Whether auto-memory features are active this session. Enabled by default;
/// `DEEPDIVE_DISABLE_AUTO_MEMORY=1` (or `true`) turns the whole subsystem off.
pub fn is_auto_memory_enabled() -> bool {
    match std::env::var("DEEPDIVE_DISABLE_AUTO_MEMORY") {
        Ok(v) => !(v == "1" || v == "true"),
        Err(_) => true,
    }
}

/// The auto-memory directory for the current project (no trailing separator).
pub fn memory_dir() -> PathBuf {
    get_project_dir().join(MEMORY_DIRNAME)
}

/// `<memoryDir>/MEMORY.md` — the index entrypoint loaded into context.
pub fn memory_entrypoint() -> PathBuf {
    memory_dir().join(ENTRYPOINT_NAME)
}

/// Ensure the memory directory exists. Idempotent, best-effort — called once per
/// session at prompt-build time so the model can Write topic files without an
/// `ls`/`mkdir` first.
pub fn ensure_memory_dir_exists() {
    let _ = std::fs::create_dir_all(memory_dir());
}

/// Is `path` inside the auto-memory directory? Used by the permission carve-out
/// (memory reads/writes never prompt) and the extraction guard (which restricts
/// the forked agent's writes to this directory).
///
/// Component-aware (`starts_with` on `Path`), so `/…/memory-other` can't slip
/// past the prefix check.
pub fn is_auto_mem_path(path: &str) -> bool {
    if path.is_empty() {
        return false;
    }
    let p = Path::new(path);
    if !p.is_absolute() {
        return false;
    }
    let dir = memory_dir();
    p == dir || p.starts_with(&dir)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_paths_under_memory_dir_only() {
        let inside = memory_dir().join("a.md");
        assert!(is_auto_mem_path(&inside.to_string_lossy()));
        assert!(is_auto_mem_path(&memory_dir().to_string_lossy()));
        assert!(!is_auto_mem_path("/etc/passwd"));
        assert!(!is_auto_mem_path("relative/path.md"));
        assert!(!is_auto_mem_path(""));
    }

    #[test]
    fn memory_display_labels_and_hides_path() {
        use crate::tools::format::memory_display;
        let mem = memory_dir().join("feedback_tests.md");
        let mem = mem.to_string_lossy();
        assert_eq!(
            memory_display("read_file", &mem),
            Some(("Recall".to_string(), "feedback_tests.md".to_string()))
        );
        assert_eq!(
            memory_display("write_file", &mem),
            Some(("Remember".to_string(), "feedback_tests.md".to_string()))
        );
        assert_eq!(
            memory_display("edit_file", &mem).map(|(d, _)| d),
            Some("Remember".to_string())
        );
        assert_eq!(memory_display("read_file", "/repo/src/a.rs"), None);
    }
}
